Add blogpost about Spyder's 5 new release (#186)
Quansight-Labs/quansight-labs-site: posts/2021/04/spot-the-differences.md

<!--
.. title: Spot the differences: what is new in Spyder 5?
.. slug: spot-the-diffenrences
.. date: 2021-04-16 08:00:00 UTC-06:00
.. author: Isabela Presedo-Floyd and Juanita Gomez
.. tags: Spyder, UX/UI, release
.. category:
.. link:
.. description:
.. type: text
-->

![Spyder 5 versus Spyder 4](/images/spyder5-header.png)

*This blog post is a conversation. Portions lead by Juanita Gomez are marked 
as **Juanita**, and those lead by Isabela Presedo-Floyd are marked as 
**Isabela**.*

> 

Reviewer: Unless folks have context this might be a bit confusing I would open with something like: 

In case you missed it Spyder 5 was released at the beginning of April and link to the repo

Reviewer: Even if this is also a few lines down - you always want to start a blog with a hook

Reviewer 2: Combining this with one the other lingering comments, I think the opening could be like

`In case you missed it Spyder 5 was released at the beginning of April! This blog post is a conversation attempting to document the long and complex process of improving Spyder's UI with this release. Portions lead by Juanita Gomez are marked  as **Juanita**, and those lead by Isabela Presedo-Floyd are marked as **Isabela**.`

and then immediately after go into the existing intro

` **[Juanita]** Spyder was created 10 years ago…`

Does that seem like it could resolve the comments and flow better?

Author: I will implement this as a separate commit once we agree on the other suggestions since it's kind of long.

## What did we do?*

> **[Juanita]** Spyder was created 10 years ago and it has had the 

Reviewer 2: ```suggestion
**[Juanita]** [Spyder](https://www.spyder-ide.org/) was created 10 years ago and it has had the 
```

contributions of a great number of developers who have written code, proposed 
ideas, opened issues and tested PRs in order to build a piece of Spyder on 

> their own. We (Spyder’s team) have been lucky to have such a great community 

Reviewer: ```suggestion
their own. We (the Spyder team) have been lucky to have such a great community 
```

of people contributing, but this is the first time that we decided to ask for 

> help from an UX/UI expert! Having the help of this great amount of people has 

Reviewer: ```suggestion
help from an UX/UI expert! Why? You might wonder. Having the help of this great amount of people has 
```

resulted in inconsistencies around Spyder’s interface which we didn’t stop to 
analyze until now. 

When Isabela joined Quansight, we realized that we had an opportunity of 
improving Spyder’s interface with her help. We thought her skill set was 

> everything we needed to make Spyder’s UI better. We conducted a community 

Reviewer: ```suggestion
everything we needed to make Spyder’s UI better. So we started by conducting a community 
```

Author: I think the survey was not actually a consequence of this, so this is not quite accurate but if it helps the flow of the post I can make the change. Thoughts @isabela-pf ?

Reviewer 2: I think it can be accurate and flow! Maybe `So we started by reviewing the results of a community survey from a few months ago and realized…`

Reviewer: My bad - but the current way it is it seems to bounce from one idea to another seemingly unrelated. I like @isabela-pf 's proposal

> everything we needed to make Spyder’s UI better. We conducted a community 

Author: ```suggestion
everything we needed to make Spyder’s UI better.  So we started by reviewing the results of a community survey
```
```suggestion
everything we needed to make Spyder’s UI better. We conducted a community 
```

> survey a couple of months ago and realized that some of the most common 

Author: ```suggestion
 from a few months ago and realized that some of the most common 
```

feedback from users is related to its interface (very crowded, not 

> consistent, many colors). This is the reason why we decided to start a joint 

Reviewer: ```suggestion
consistent, many colors). This is why we decided to start a joint 
```

project with Isabela, (who we consider now part of Spyder’s team) called 

> Spyder 5!!!

Reviewer 2: ```suggestion
[Spyder 5](https://github.com/spyder-ide/spyder/releases/tag/v5.0.0)!!!
```

> project with Isabela, (who we consider now part of Spyder’s team) called 

Reviewer: ```suggestion
project with Isabela, (who we consider now part of the Spyder team) called 
```

> 

Reviewer: I would add something like - this blogpost is an attempt to document the long and complex process of improving Spyder's UI. To give this a nice intro.

Author: I like the idea of adding this but maybe at the beginning along with the mention of the release? Cause I think in between these two paragraphs it interrupts a bit the flow since the next paragraph starts with the "version" of spyder which was mentioned at the end of the paragraph before. What do you think @trallard and @isabela-pf ?

Reviewer 2: I could either option, but maybe it makes the most sense wherever the `*This blog post is a conversation…*` goes. Both are meant to give the post context and explain the motivation, so I would keep them together.

<!-- TEASER_END -->

This version was in development for over a year and was finally released on 
April 1st. It has great new features that we hope will benefit our users 
greatly. Most of these are focused on improving Spyder’s interface and 
usability, which we did thanks to Isabela’s help. The 3 main UX features 
implemented in this release were:

- A brand new color palette designed to bring greater consistency to the UI 
and to make it easier to use.

> - The redesign of our toolbars by adjusting the margins and sizes of all the 

Author: I think "A redesign" sounds very weird.

buttons to meet accessibility recommendations.
- A new set of icons to ensure a consistent style.

> ##How did we do it? (conversation)

Reviewer 2: ```suggestion
## How did we do it?
```

back. Because I have a habit of pushing every button in an interface, I ended 
up having a lot (maybe too much) feedback to pass on. One of the things I 
remember most about opening Spyder for the first time was having three dialogs 
pop up immediately. That’s really not the first impression you want to give, 
and I remember talking to Juanita about that right away. Figuring out how to 
state problems as simply and clearly to a group of people I didn’t know yet 
was intimidating and went through several phases.

### 2. From the “nightmare document” to the issue tracker

**[Juanita]** The first phase was discussing all the problems that Isabela 
found in weekly meetings with Carlos, the Spyder maintainer, and Stephanie, 
a core developer for Spyder. I created a Google drive document (which we 
ended up calling “The Nightmare document”) in which I collected most of the 
feedback that Isabela gave us. Then, I grouped this information into 
categories depending on whether the comments were about the interface, in 
general, or if they were about a specific pane. Once we agreed on a relevant 
problem that we wanted to address, I opened an issue on a new repo that we 

> created in the Spyder’s organization called “ux-improvements.”

Reviewer 2: ```suggestion
created in the Spyder’s organization called “[ux-improvements](https://github.com/spyder-ide/ux-improvements/issues).”
```

**[Isabela]** In fact, that wasn’t even the first set of documents we were 
working with; I had a whole table, numbering system, and document I was 
trying to handle before. But it was Juanita that turned them into Github 
issues.

### 3. Sorting out the nightmare 

**[Juanita]** Since we ended up with more than 30 issues, we had to start a 
“triaging phase.” We had to label, triage, organize, and prioritize issues 
according to “urgency” and importance. This issue tracker became our main 
tool to keep up with all the plans for the future!

**[Isabela]** Juanita did wonderful work tracking our progress through issues 
and keeping us all accountable, but we were still left with a long list of 
issues to triage—long enough that it wasn’t all getting in Spyder 5. To have 

> the greatest impact on the Spyder, we started with the issues that had 

Reviewer: ```suggestion
the greatest impact on Spyder, we started with the issues that had 
```

influence on Spyder as a whole. Toolbars, icons, and colors are something y
ou will always encounter from the first impression to the most recent, so it 
made sense to start thinking about those big picture issues first.

### 4. Digging deeper into the dark hole

**[Isabela]** When prioritizing the audit feedback for Spyder 5, each pass 
seemed to get to a deeper layer of the problem. For example, what started as 

> issues to make tooltips more legible and improve the variable explorer’s 

Reviewer 2: ```suggestion
issues to make [tooltips more legible](https://github.com/spyder-ide/ux-improvements/issues/2) and improve the variable explorer’s 
```

> color coding soon became the realization that we weren’t sure exactly what 

Reviewer 2: ```suggestion
[color coding](https://github.com/spyder-ide/ux-improvements/issues/7) soon became the realization that we weren’t sure exactly what 
```

blue was already being used for much of Spyder’s interface. It got more 
complicated when we found out how many colors were hard coded across multiple 
files or defined by an external project. Eventually, the problem changed from 

> the color contrast of tool tips to an unsustainable approach for managing 

Reviewer 2: ```suggestion
the color contrast of tool tips to an [unsustainable approach for managing 
```

> color across the two default Spyder themes rooted in a non-Spyder repo. Work 

Reviewer 2: ```suggestion
color](https://github.com/spyder-ide/ux-improvements/issues/13) across the two default Spyder themes rooted in a non-Spyder repo. Work 
```

at each step did build up into a larger solution, but it’s worth noting that 
it isn’t what we set out to do in the first place. 

### 5. What witchcraft does Isabela do in the background?

**[Juanita]** One of the most important parts of the process was designing 
the mock ups for the new ideas that we came up with for the interface which  
is definitely not our expertise. So... how did the designs magically appear 
on our Github issues?

**[Isabela]** First things first, it isn’t actually witchcraft even if it 
looks magical from the outside. How I work depends somewhat on what problem 

> e are trying to solve, so let’s use the design of custom icons for Spyder 5 

Reviewer 2: ```suggestion
e are trying to solve, so let’s use the design of [custom icons](https://github.com/spyder-ide/ux-improvements/issues/33#issuecomment-776376943) for Spyder 5 
```

as an example. Once I had a defined list of icons to work on, I needed to 
spend time making progress on my own. Research on best practices for the 
relevant area of design is where I started; in this case, I knew we were 

> going to be working with Material Design Icons’ specifications. After that, 

Reviewer 2: ```suggestion
going to be working with [Material Design Icons’](https://materialdesignicons.com/) specifications. After that, 
```

I did a combination of pen-and-paper sketching and working digitally based on 

> the existing icons in Spyder and Material Design Icons while I kept note of 

Author: I @isabela-pf wrote it like this because it is important to make clear that the distribution is called "Material Design Icons" from https://materialdesignicons.com and not "Material Design" from https://material.io/design.

the pros and cons for different directions. I also collected design elements 
as I built them so that I could make more consistent, accurate designs faster 
as I worked. For the icons, this included things like letters, rounded 
corners, and templates for the size and spacing of elements. Finally, I 
compared options side by side and tried them out in the interface to evaluate 
what designs were strong enough to bring to the rest of the team. Then we 
discussed the options together.

![Spyder 5 icon mockups](/images/spyder5-wipicons.png)

### 6. Mockups Vs Reality

**[Juanita]** After many discussions, mock-ups, and meetings, decisions had 
been made. It was time to implement! A big part of the improvements were made 

> in QDarkStyleSheet where we did the new palette and color system for both the 

Reviewer 2: ```suggestion
in [QDarkStyleSheet](https://github.com/ColinDuquesnoy/QDarkStyleSheet/) where we did the new palette and color system for both the 
```

> what designs were strong enough to bring to the rest of the team. Then we 

Author: There is a "them" in the previous sentence so I don't think it is necessary. But since @isabela-pf wrote this, I'll let her decide.

> what designs were strong enough to bring to the rest of the team. Then we 

Reviewer 2: I'm happy with this suggestion; thanks for asking!

> **[Juanita]** After many discussions, mock-ups, and meetings, decisions had 

Reviewer: ```suggestion
**[Juanita]** After many discussions, mock-ups, and meetings, decisions were
```

> been made. It was time to implement! A big part of the improvements were made 

Reviewer: ```suggestion
made and we were ready to move onto the implementation phase. A big part of the improvements were made 
```

dark and light themes of Spyder. In my opinion, this was the hardest part of 
the process since it involved getting familiar with the code first and then, 
trying and trying again changing lines of code to change the color or style 
of buttons, tabs, toolbars, borders, etc… 

The other problem that I ran into, was trying to meet the designs’ 

> specifications. Specially, when working with the toolbars, figuring the right 

Reviewer 2: ```suggestion
specifications. Specially, when working with the [toolbars](https://github.com/spyder-ide/ux-improvements/issues/28), figuring the right 
```

number for the pixels of margins and sizes was a challenge. I tried several 
values before finding one that closely matched the proposed mock-up only to 
realize later that “pixels” was not the best unit for the specifications. I 
ended up using “em” since it was more consistent across operating systems.
Isabela, Stephanie and Carlos were part of this process as well. Between the 
four of us we managed to implement all the changes that we had planned for 
Spyder 5, the new color palette, the complete redesign of toolbars and the 
new set of icons. It was an arduous task, more than we all expected, but at 
the end we were all very happy with the results and thankful to Isabela for 
helping us to give a new face to Spyder. 

> ## What did we get?

Reviewer: Maybe: The final result

Like a reveal

Author: Not quite sure about this one since we are trying to make all the section headers be questions. What do you think @isabela-pf?

Reviewer 2: I do think it might stand out to have only one header in this section not be a question. Would `What's the final result?` work?

> ## What did we get?

Author: ```suggestion
## What's the final result?
```

interface; they are more noticeable, and rooted in the Spyder UI many people 
are already familiar with. While the changes may feel obvious when they are 
new, they are also chosen to create consistent patterns across interactions 
that can become more comfortable over time. Spyder’s default dark and light 
modes, for example, used to use a different set of UI elements between modes. 
Now they both use the same elements and it is only the colors that change. 
This makes it easier for users to jump into a familiar interface and take 
what they know from working in one space to another. For contributors, it 
gives a more clear UI pattern for them to follow in their own work.

### Before and after (Dark theme)

![Spyder 5 vs Spyder 4 dark theme](/images/spyder5-comparedark.png)

### Before and after (Light theme)

![Spyder 5 vs Spyder 4 light theme](/images/spyder5-comparelight.png)

## What did we learn :)?

**[Isabela]** From developing new skills to working as a team for the first 
time, I think we both took a lot from this process. Here are some lessons that 
stood out to us.

**[Juanita]** 

- Sometimes it is better to try some of the ideas during the process, than 

> having long discussions about an idea and implementing at the end. In some 

Reviewer 3: ```suggestion
having long discussions about an idea and implementing it at the end. In some 
```

cases you end up realizing that things don’t look as good as you thought they 
would, or that some are not even possible.
- One of the most important parts of the design process is to get yourself in 
the users’ shoes. At the end, they are the reason why we work to improve 
things constantly.
- Occasionally, less is more. Simple and consistent is better than crowded 
and complicated. 

**[Isabela]** 

- Don’t be afraid of asking questions even when you think you understand the 

> problem because every bit of information can be useful to understanding what 

Reviewer 3: ```suggestion
problem because every bit of information can be useful to better grasping what 
```

Avoids saying `understand` and `understanding` twice in the same sentence.

<!--
.. title: Spot the differences: what is new in Spyder 5?
.. slug: spot-the-diffenrences
.. date: 2021-04-16 08:00:00 UTC-06:00
.. author: Isabela Presedo-Floyd and Juanita Gomez
.. tags: Spyder, UX/UI, release
.. category:
.. link:
.. description:
.. type: text
-->

![Spyder 5 versus Spyder 4](/images/spyder5-header.png)

In case you missed it, Spyder 5 was released at the beginning of April! This 
blog post is a conversation attempting to document the long and complex 
process of improving Spyder's UI with this release. Portions lead by Juanita 
Gomez are marked as **Juanita**, and those lead by Isabela Presedo-Floyd are 
marked as **Isabela**.

## What did we do?*

**[Juanita]** [Spyder](https://www.spyder-ide.org/) was created more than 10 

> years ago and it has had the  contributions of a great number of developers 

Reviewer 2: ```suggestion
years ago and it has had the contributions of a great number of developers 
```

of buttons, tabs, toolbars, borders, etc… 

The other problem that I ran into, was trying to meet the designs’ 
specifications. Specially, when working with the [toolbars](https://github.com/spyder-ide/ux-improvements/issues/28), figuring the right 
number for the pixels of margins and sizes was a challenge. I tried several 
values before finding one that closely matched the proposed mock up only to 
realize later that “pixels” was not the best unit for the specifications. I 
ended up using “em” since it was more consistent across operating systems.
Isabela, Stephanie and Carlos were part of this process as well. Between the 
four of us we managed to implement all the changes that we had planned for 
Spyder 5, the new color palette, the complete redesign of toolbars and the 
new set of icons. It was an arduous task, more than we all expected, but at 
the end we were all very happy with the results and thankful to Isabela for 
helping us to give a new face to Spyder. 

## What's the final result?

**[Isabela]** Individually, the colors, toolbars, and icons may feel like 
small adjustments, but those are some of the elements that make up most of 

> Spyder.  When they are together, those small adjustments set the mood in the 

Reviewer 2: ```suggestion
Spyder. When they are together, those small adjustments set the mood in the 
```
